### PR TITLE
Rework vulkan buffer allocation

### PR DESCRIPTION
**Pull request by 0lvin**
* replace vk buffer allocation with code from q2vkpt/remove c++ code #505
* fix screenshot creation
* remove non necessary scale image with retexturing

### CODE REVIEW

**Comment by Yamagi** [comment]
Thank you for you continued work. It would be nice if anybody with somewhat more Vulkan knowledge than @DanielGibson or me could have a look at this. All I can say is that it seems to work. :)

File: Makefile [hunk 1]
```diff
@@ -99,7 +99,6 @@ endif
 # On Windows / MinGW $(CC) is undefined by default.
 ifeq ($(YQ2_OSTYPE),Windows)
 CC := gcc
-CXX := g++
```
Comment (turol) [inline Makefile]: The Makefile changes should be their own separate commit after everything else.

File: src/client/refresh/vk/header/vk_mem_alloc.h [hunk 2]
```diff
@@ -1,17087 +0,0 @@
-//
```
Comment (turol) [inline src/client/refresh/vk/header/vk_mem_alloc.h]: Removing this file should be its own separate commit so it doesn't clutter up this one

File: src/client/refresh/vk/vk_buffer.c [hunk 3]
```diff
@@ -38,33 +38,27 @@ static void copyBuffer(const VkBuffer *src, VkBuffer *dst, VkDeviceSize size)
 }
 
 // internal helper
-static void createStagedBuffer(const void *data, VkDeviceSize size, qvkbuffer_t *dstBuffer, qvkbufferopts_t bufferOpts, qvkbuffer_t *stagingBuffer)
+static void createStagedBuffer(const void *data, VkDeviceSize size, qvkbuffer_t *dstBuffer, qvkbufferopts_t bufferOpts)
```
Comment (turol) [inline src/client/refresh/vk/vk_buffer.c]: You've mixed formatting changes with functional changes. This makes it extremely hard to check what was actually changed.

File: src/client/refresh/vk/vk_image.c [hunk 4]
```diff
@@ -336,49 +336,32 @@ VkResult QVk_CreateImage(uint32_t width, uint32_t height, VkFormat format, VkIma
...
 	texture->sharingMode = imageInfo.sharingMode;
-	return vmaCreateImage(vk_malloc, &imageInfo, &vmallocInfo, &texture->image, &texture->allocation, &texture->allocInfo);
+	return image_create(&texture->resource, imageInfo, /*mem_properties*/ 0, /*mem_preferences*/ 0);
 }
 
 void QVk_CreateDepthBuffer(VkSampleCountFlagBits sampleCount, qvktexture_t *depthBuffer)
 {
 	depthBuffer->format = QVk_FindDepthFormat();
 	depthBuffer->sampleCount = sampleCount;
-	// On 64-bit builds, Intel drivers throw a warning:
```
Comment (turol) [inline src/client/refresh/vk/vk_image.c]: Did you check on an Intel GPU that this issue is not present with your code?

File: src/client/refresh/vk/header/vk_local.h [hunk 5]
```diff
@@ -227,7 +227,9 @@ qboolean	R_EndWorldRenderReady( void );
 struct image_s *R_RegisterSkin (char *name);
 
 void LoadPCX (char *filename, byte **pic, byte **palette, int *width, int *height);
-image_t *Vk_LoadPic (char *name, byte *pic, int width, int height, imagetype_t type, int bits, qvksampler_t *samplerType);
+image_t *Vk_LoadPic(char *name, byte *pic, int width, int realwidth,
```
Comment (turol) [inline src/client/refresh/vk/header/vk_local.h]: More mixed functional and formatting changes.

File: src/client/refresh/vk/vk_util.c [hunk 6]
```diff
@@ -210,7 +210,7 @@ buffer_flush(BufferResource_t *buf)
 	VkResult result = VK_SUCCESS;
 
 	VkMappedMemoryRange ranges[1] = {{
-		.sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
+		.sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE,
```
Comment (turol) [inline src/client/refresh/vk/vk_util.c]: This doesn't look like debug code. This commit should be merged with one of the previous ones so there are no broken commits in case someone has to bisect.

**Comment by 0lvin** [comment]
> On what GPUs and platforms was this tested? Does it pass validation cleanly?

Code have tested on RADV and SwiftShader with enabled _DEBUG and VK_LAYER_KHRONOS_validation.

File: Makefile (hunk 1 above)
Comment (0lvin) [reply Makefile]: thank you fixed in 6f3c223

File: src/client/refresh/vk/header/vk_mem_alloc.h (hunk 2 above)
Comment (0lvin) [reply src/client/refresh/vk/header/vk_mem_alloc.h]: thank you, fixed in 6f3c223

File: src/client/refresh/vk/vk_buffer.c (hunk 3 above)
Comment (0lvin) [reply src/client/refresh/vk/vk_buffer.c]: thank you, all codestyle changes moved to 695a9e9

File: src/client/refresh/vk/vk_image.c (hunk 4 above)
Comment (0lvin) [reply src/client/refresh/vk/vk_image.c]: I don't have Intel hardware, could some one check it?

File: src/client/refresh/vk/header/vk_local.h (hunk 5 above)
Comment (0lvin) [reply src/client/refresh/vk/header/vk_local.h]: thank you, all codestyle changes moved to 695a9e9

File: src/client/refresh/vk/vk_util.c (hunk 6 above)
Comment (0lvin) [reply src/client/refresh/vk/vk_util.c]: thank you, I have fix to original commit  3dc4fa0

**Comment by 0lvin** [comment]
@turol I have fixed code and split code with replace new allocator to C implementation.

Could you please look?

**Comment by turol** [comment]
Vulkan-wise it looks ok. Using individual allocation per buffer/image is suboptimal but shouldn't break anything in this case. You've still got some unrelated whitespace changes mixed with functional changes. Also there are typos in commit messages and comments. It's up to someone else if those are blocking issues.

**Comment by Yamagi** [comment]
Can I merge this in the feature branch?

**Comment by 0lvin** [comment]
Yes, it will great. Thank you.

**Comment by Yamagi** [comment]
Cool. And merged. :)
